### PR TITLE
String listfield bugfix

### PR DESCRIPTION
**Pull request by adamalix**
Refactoring, finding an item in a listfield and doc strings.

### CODE REVIEW

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 1]
```diff
@@ -1,109 +1,173 @@
-// Copyright 2011 Foursquare Labs Inc. All Rights Reserved.
+// Copyright 2012 Foursquare Labs Inc. All Rights Reserved.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Make 2011-2012

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 2]
```diff
@@ -1,109 +1,173 @@
...
+   */
+  def quote(q: String): String = "\"" + q + "\""
 
+  /**
+   * AbstractClause represents query clauses.
+   */
   abstract class AbstractClause {
-    def extend: String
+    /**
+     * Returns the clause's string representation
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Maybe mention Solr  in this

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 3]
```diff
@@ -120,88 +186,138 @@ object Ast {
...
       "(doc['" + fieldName + "'].value)"
     }
   }
 
-  //A field with a query weight
-  case class WeightedField(fieldName: String, boost: Double = 1) extends ScoreBoost {
-    def extend(): String = {
-      boost match {
+  /**
+   * Class representing a field that can have its score boosted by a weight.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: This class represents a boost (multiplier) for matches occurring in the given field.

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 4]
```diff
@@ -120,88 +186,138 @@ object Ast {
...
-      new QueryFilterBuilder(this.elasticExtend(qf, Nil, None))
+      EFilterBuilders.queryFilter(this.elasticExtend(qf, Nil, None))
     }
-
   }
 
-  //Name doesn't have to be a field name for Solr
-  //it could be "lat,lng". However for ES it must be
-  //a field
+  /**
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala:297]: This comment belongs down below with the GeoDist class

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 5]
```diff
@@ -209,103 +325,138 @@ object Ast {
...
   case class Empty[T]() extends Query[T] {
-    def extend = "\"\""
+    /** @inheritdoc */
+    def extend(): String = "\"\""
+    /** @inheritdoc */
     def elasticExtend(qf: List[WeightedField], pf: List[PhraseWeightedField], mm: Option[String]): ElasticQueryBuilder = {
-      val q = new QueryStringQueryBuilder(this.extend)
-      qf.map(f => q.field(f.fieldName,f.boost.toFloat))
+      val q = EQueryBuilders.queryString(this.extend())
+      qf.map(f => q.field(f.fieldName,f.weight.toFloat))
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala:350]: We can probably kill this line? Lets add a test case and see if we can (after the merge)

File: src/main/scala/com/foursquare/slashem/QueryBuilder.scala [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2011 Foursquare Labs Inc. All Rights Reserved.
+// Copyright 2012 Foursquare Labs Inc. All Rights Reserved.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/QueryBuilder.scala]: s/2012/2011-2012/

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 7]
```diff
@@ -1,53 +1,54 @@
-// Copyright 2011 Foursquare Labs Inc. All Rights Reserved.
+// Copyright 2012 Foursquare Labs Inc. All Rights Reserved.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: s/2012/2011-2012/

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 8]
```diff
@@ -56,77 +57,95 @@ case class SolrResponseException(code: Int, reason: String, solrName: String, qu
...
                                        numFound: Int, start: Int, docs: Array[Response.RawDoc],
-                                       fallOf: Option[Double], min: Option[Int], fieldFacets: Map[String,Map[String,Int]]) {
-  val filteredDocs = filterHighQuality(docs)
+                                       fallOff: Option[Double], min: Option[Int],
+                                       fieldFacets: Map[String,Map[String,Int]]) {
+  // Docs with high Lucene scores.
+  val filteredDocs: Array[Response.RawDoc] = filterHighQuality(docs)
+
+  /**
+   * Gets a List[T] of high scoring docs returned from Lucene.
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: This comment is misleading (its only filtered to high scoring under certain circumstances), remove.

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 9]
```diff
@@ -120,88 +186,138 @@ object Ast {
...
+     */
     def or(c: Query[T]): Query[T] = Or(this, c)
-    def boost(b: Float): Query[T] = Boost(this, b)
+    /**
+     * Boost a query by a weight
+     * @see Boost
+     */
+    def boost(weight: Float): Query[T] = Boost(this, weight)
+    /**
+     * @return String Query's string representation
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Its not really the query's string representation, its Solr's query format

File: src/main/scala/com/foursquare/slashem/Ast.scala (hunk 1 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Done.

File: src/main/scala/com/foursquare/slashem/Ast.scala (hunk 2 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Done.

File: src/main/scala/com/foursquare/slashem/Ast.scala (hunk 3 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Done.

File: src/main/scala/com/foursquare/slashem/Ast.scala (hunk 9 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Ast.scala]: Done.

File: src/main/scala/com/foursquare/slashem/Ast.scala (hunk 4 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Ast.scala:297]: Done.

File: src/main/scala/com/foursquare/slashem/Ast.scala (hunk 5 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Ast.scala:350]: Will do.

File: src/main/scala/com/foursquare/slashem/QueryBuilder.scala (hunk 6 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/QueryBuilder.scala]: Done.

File: src/main/scala/com/foursquare/slashem/Schema.scala (hunk 7 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: Done.

File: src/main/scala/com/foursquare/slashem/Schema.scala (hunk 8 above)
Comment (adamalix) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: Changed to:
- Gets a List[T] of docs returned from Lucene.

File: src/main/scala/com/foursquare/slashem/Ast.scala [hunk 10]
```diff
@@ -209,103 +327,138 @@ object Ast {
...
       q
     }
   }
 
+  /**
+   * A phrase containing a query that is optionally escaped.
+   *
+   * Represents a contiguous series of words to be matched in that order.
+   */
   case class Phrase[T](query: T, escaped: Boolean = true) extends Query[T] {
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Ast.scala:360]: TODO: Listen to escaped (post-merge)

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 11]
```diff
@@ -293,13 +319,16 @@ trait SolrMeta[T <: Record[T]] extends SlashemMeta[T] {
         Pair(doc.toMap,hl)})
       val facetCounts = rsr.facetCounts
 
-      val facets: scala.collection.immutable.Map[String,scala.collection.immutable.Map[String,Int]] = if (facetCounts != null) {
-        facetCounts.facetFields.asScala.map(magic => (magic._1,parseFacetCounts(magic._2.asScala.toList).toMap)).toMap
+      val facets: Map[String,Map[String,Int]] = if (facetCounts != null) {
+        facetCounts.facetFields.asScala.map(magic => {
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: As discussed, if you want a better name for magic (for some _crazy_ reason :p), go with ffCountPair

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 12]
```diff
@@ -341,7 +366,8 @@ trait SolrMeta[T <: Record[T]] extends SlashemMeta[T] {
   }
 
 }
-//If you want to get some simple logging/timing implement this trait
+
+/** Provides simple logging/timing */
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: This isn't really true, the trait doesn't provide anything

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 13]
```diff
@@ -545,12 +571,12 @@ trait ElasticSchema[M <: Record[M]] extends SlashemSchema[M] {
...
   }
   def termFacetQuery(facetFields: List[Ast.Field], facetLimit: Option[Int]): List[AbstractFacetBuilder] = {
-    val fieldNames = facetFields.map(_.extend())
+    val fieldNames = facetFields.map(_.boost())
     val facetQueries = fieldNames.map(name => {
       val q = new TermsFacetBuilder(name).field(name)
       facetLimit match {
         case Some(c) => {
-          println("using size of "+c);
+          println("using size of "+c)
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: kill this println, I should not have committed this

File: src/main/scala/com/foursquare/slashem/Schema.scala [hunk 14]
```diff
@@ -795,6 +823,27 @@ class SlashemStringListField[T <: Record[T]](owner: T) extends StringListField[T
...
+}
+
+class SlashemLongListField[T <: Record[T]](owner: T) extends LongListField[T](owner) with SlashemField[List[Long], T] {
+  override def valueBoxFromAny(a: Any) = {
+    try {
+      a match {
+        case long:       Long => Full(List(long))
+        case strArr:     Array[Long]  => Full(strArr.toList)
+        case intArr:     Array[Int]     => Full(intArr.toList.map(int => int.toLong))
+        case _ => Empty
```
Comment (holdenk) [inline src/main/scala/com/foursquare/slashem/Schema.scala]: What about the String case (i.e. when being returned from Solr?)

**Pull request merged**